src/ui/camera.js

```diff
@@ -632,66 +747,92 @@ class Camera extends Evented {
-        const p3rotated = p3world.rotate(angleRadians);
-
-        const upperRight = new Point(
-            Math.max(p0rotated.x, p1rotated.x, p2rotated.x, p3rotated.x),
-            Math.max(p0rotated.y, p1rotated.y, p2rotated.y, p3rotated.y)
-        );
-        const lowerLeft = new Point(
-            Math.min(p0rotated.x, p1rotated.x, p2rotated.x, p3rotated.x),
-            Math.min(p0rotated.y, p1rotated.y, p2rotated.y, p3rotated.y)
-        );
-
-        // Calculate zoom: consider the original bbox and padding.
-        const size = upperRight.sub(lowerLeft);
-        const scaleX = (tr.width - ((edgePadding.left || 0) + (edgePadding.right || 0) + eOptions.padding.left + eOptions.padding.right)) / size.x;
-        const scaleY = (tr.height - ((edgePadding.top || 0) + (edgePadding.bottom || 0) + eOptions.padding.top + eOptions.padding.bottom)) / size.y;
-
-        if (scaleY < 0 || scaleX < 0) {
-            warnOnce(
-                'Map cannot fit within canvas with the given bounds, padding, and/or offset.'
-            );
-            return;
-        }
-        const zoom = Math.min(tr.scaleZoom(tr.scale * Math.min(scaleX, scaleY)), eOptions.maxZoom);
 
-        // Calculate center: apply the zoom, the configured offset, as well as offset that exists as a result of padding.
-        const offset = (typeof eOptions.offset.x === 'number' && typeof eOptions.offset.y === 'number') ?
+        const worldCoords = [[p0world.x, p0world.y, 0], [p1world.x, p1world.y, 0]];
+        let aabb = Aabb.fromPoints(worldCoords);
+
+        const worldToCamera = tr.getWorldToCameraMatrix();
+        const cameraToWorld = mat4.invert(new Float64Array(16), worldToCamera);
+
+        aabb = Aabb.applyTransform(aabb, worldToCamera);
+
+        const size = vec3.sub([], aabb.max, aabb.min);
+
+        const screenPadL = edgePadding.left || 0;
+        const screenPadR = edgePadding.right || 0;
+        const screenPadB = edgePadding.bottom || 0;
+        const screenPadT = edgePadding.top || 0;
+
+        const {left: padL, right: padR, top: padT, bottom: padB} = eOptions.padding;
+
+        const halfScreenPadX = (screenPadL + screenPadR) * 0.5;
+        const halfScreenPadY = (screenPadT + screenPadB) * 0.5;
+
+        const scaleX = (tr.width - (screenPadL + screenPadR + padL + padR)) / size[0];
+        const scaleY = (tr.height - (screenPadB + screenPadT + padB + padT)) / size[1];
+
+        const zoomRef = Math.min(tr.scaleZoom(tr.scale * Math.min(scaleX, scaleY)), eOptions.maxZoom);
+
+        const scaleRatio = tr.scale / tr.zoomScale(zoomRef);
+
+        aabb = new Aabb(
+            [aabb.min[0] - (padL + halfScreenPadX) * scaleRatio, aabb.min[1] - (padB + halfScreenPadY) * scaleRatio, aabb.min[2]],
+            [aabb.max[0] + (padR + halfScreenPadX) * scaleRatio, aabb.max[1] + (padT + halfScreenPadY) * scaleRatio, aabb.max[2]]);
+
+        const aabbHalfExtentZ = (aabb.max[2] - aabb.min[2]) * 0.5;
```

Comment:
Nit: `z` doesn't change:

```suggestion
        const aabbHalfExtentZ = size[2] * 0.5;
```